File: Pythoncode/322.零钱兑换.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = {}
        res = self.helper(coins, amount, dp)
        return -1 if res == float("inf") else res 

    def helper(self, coins, amount, dp):
        if amount == 0:
            return 0
        if amount  < 0:
            return float("inf")
        if amount in dp:
            return dp[amount]
        dp[amount] = min([self.helper(coins, amount - c, dp) for c in coins]) + 1
        return dp[amount]

File: Cppcode/322.零钱兑换.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, -1);
        dp[0] = 1;
        for (auto i = 1; i < amount+1; i++)
        {
            for (auto c: coins)
            {
                if (i-c < 0 || dp[i-c] == -1) continue;
                if (dp[i] == -1) dp[i] = dp[i-c] + 1;
                else dp[i] = min(dp[i], dp[i-c] + 1);
            }
        }
        return dp.back();
    }
};